File: Dungeon Game.py
class Solution:
	# @param dungeon, a list of lists of integers
	# @return a integer
	def calculateMinimumHP(self, dungeon):
		M = len(dungeon)
		N = len(dungeon[0])
		minimumHPBeforeStepIn = [ [0]*N ] * M
		for row in range(M-1, -1, -1):
			for col in range(N-1, -1, -1):
				right = 9999999
				if (row < M - 1):
					right = minimumHPBeforeStepIn[row + 1][col]
				bottom = 9999999
				if (col < N - 1):
					bottom = minimumHPBeforeStepIn[row][col + 1]
				if (row == M - 1 and col == N - 1):
					right = bottom = 1
				# minHP > 0; minHP + dungeon > min(right, bottom)
				minimumHPBeforeStepIn[row][col] = max(1, min(right, bottom) - dungeon[row][col])
		return minimumHPBeforeStepIn[0][0]

File: Dungeon Game.cpp
class Solution {
public:
    int calculateMinimumHP(vector<vector<int> > &dungeon) {
        int M = dungeon.size();
        int N = dungeon[0].size();
        vector<vector<int> > minHPbeforeStepIn = dungeon;
        for (int r = M-1; r >= 0; r--) {
            for (int c = N-1; c >= 0; c--) {
                int right = INT_MAX;
                if (c < N - 1)
                    right = minHPbeforeStepIn[r][c+1];
                int bottom = INT_MAX;
                if (r < M - 1)
                    bottom = minHPbeforeStepIn[r+1][c];
 				if (r == M-1 && c == N-1)
				{
					right = bottom = 1;	// destination
				}
				// minHP >= 1 && (minHP + dungeon >= right || minHP + dungeon >= bottom)
                minHPbeforeStepIn[r][c] = max(1, min(right, bottom) - dungeon[r][c]);
            }
        }
        return minHPbeforeStepIn[0][0];
    }
};
